cupy/lib/_routines_poly.py

```diff
@@ -103,6 +103,65 @@ def polymul(a1, a2):
     return cupy.convolve(a1, a2)
 
 
+def polyval(p, x):
+    """Evaluates a polynomial at specific values.
+
+    Args:
+        p (cupy.ndarray or cupy.poly1d): input polynomial.
+        x (scalar, cupy.ndarray): values at which the polynomial
+        is evaluated.
+
+    Returns:
+        cupy.ndarray or cupy.poly1d: polynomial evaluated at x.
+
+    .. warning::
+
+        This function doesn't currently support poly1d values to evaluate.
+
+    .. seealso:: :func:`numpy.polyval`
+
+    """
+    if isinstance(p, cupy.poly1d):
+        p = p.coeffs
+    if not isinstance(p, cupy.ndarray) or p.ndim == 0:
+        raise TypeError('p can be 1d ndarray or poly1d object only')
+    if p.ndim != 1:
+        # to be consistent with polyarithmetic routines' behavior of
+        # not allowing multidimensional polynomial inputs.
+        raise ValueError('p can be 1d ndarray or poly1d object only')
+        # TODO(Dahlia-Chehata): Support poly1d x
+    if (isinstance(x, cupy.ndarray) and x.ndim <= 1) or numpy.isscalar(x):
+        val = cupy.asarray(x).reshape(-1, 1)
+    else:
+        raise NotImplementedError(
+            'poly1d or non 1d values are not currently supported')
+    out = p[::-1] * cupy.power(val, cupy.arange(p.size))
+    out = out.sum(axis=1)
+    dtype = cupy.result_type(p, val)
+    # For case: when p is of shape (0,) and x is (), output is
+    #  of single valued array to match NumPy's behavior
+    if isinstance(x, cupy.ndarray) and x.ndim == 0 and p.size == 0:
+        return out[0].astype(dtype, copy=False)
```

Comment:
Why is this case needed? `cupy.isscalar(x) or x.ndim == 0` in L146 seems to cover this case.